socs/agents/scpi_psu/drivers.py

```diff
@@ -1,6 +1,150 @@
 # Tucker Elleflot
+
+import socket
+import time
+
 from socs.common.prologix_interface import PrologixInterface
 
+# append new model strings as needed
+ONE_CHANNEL_MODELS = ['2280S-60-3', '2280S-32-6']
+THREE_CHANNEL_MODELS = ['2230G-30-1']
+
+
+class ScpiPsuInterface:
+    def __init__(self, ip_address, gpibAddr, port, **kwargs):
+        self.ip_address = ip_address
+        self.gpibAddr = gpibAddr
+        self.port = port
+        self.sock = None
+        self.model = None
+        self.numChannels = 0
+        self.conn_socket()
+        try:
+            self.configure()
+        except ValueError as err:
+            raise ValueError(err)
+        super().__init__(**kwargs)
+
+    def conn_socket(self):
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.sock.connect((self.ip_address, self.port))
+        self.sock.settimeout(5)
+
+    def read(self):
+        return self.sock.recv(128).decode().strip()
+
+    def write(self, msg):
+        message = msg + '\n'
+        self.sock.sendall(message.encode())
+        time.sleep(0.1)  # to prevent flooding the connection
+
+    def identify(self):
+        self.write('*idn?')
+        return self.read()
+
+    def read_model(self):
+        idn_response = self.identify().split(',')[1]
+        if (idn_response.startswith('MODEL')):
+            return idn_response[6:]
+        else:
+            return idn_response
+
+    def configure(self):
+        self.model = self.read_model()
+        if (self.model in ONE_CHANNEL_MODELS):
+            self.numChannels = 1
+        if (self.model in THREE_CHANNEL_MODELS):
+            self.numChannels = 3
+        if (self.numChannels == 0):
+            raise ValueError('Model number not found in known device models', self.model)
```

Comment:
agreed, the reason I eventually made a whole separate class for this was honestly just laziness; I didn't want to deal with the PrologixInterface inheritance. In the agent code I choose which class to use based on the presence of the 'port' argument; if this driver class is chosen the ValueError is caught and the error text is logged.

It would probably be better to query the device for number of channels rather than maintaining a hard-coded list of model numbers. For now, should I just change the error message to a warning and set numChannels to 3 as the default?

Reply:
Yeah, that sounds good to me. And so does querying for number of channels.

Reply:
I did a dive into querying for number of channels, and I can't see anyway to do it with the universally implemented SCPI commands. Each device and manufacturer implements different commands. It seems like the only way to query for number of channels would be to iterate through a set of functions, one for each device, and then check the error flags. Of course this would be impractical to implement for all possible SCPI power supplies that might be in use.

It's unfortunate that the set of required commands for SCPI do not include some way of checking for number of channels. I've retained the lists of known model numbers for now, changed the logged error into a logged warning and set num_channels to 3 if the model number is not found.

Reply:
That is unfortunate, and something I don't think we realized when we renamed this agent from "Keithley" to "SCPI PSU" agent way back when. Thanks for digging into the details though! Keithley PSUs are the main use case here, with one or two BK Precision PSUs I believe also being used. Though I don't have a good idea of how regularly this agent is used these days. That said, I wouldn't recommend spending more time generalizing this unless the functionality is in demand.